File: java/core/src/test/java/com/smartness/traceflow/handles/StepHandleTest.java
package com.smartness.traceflow.handles;

import com.smartness.traceflow.dto.TraceEvent;
import com.smartness.traceflow.enums.LogLevel;
import com.smartness.traceflow.enums.TraceEventType;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class StepHandleTest {

    private List<TraceEvent> events;
    private StepHandle handle;

    @BeforeEach
    void setUp() {
        events = new ArrayList<>();
        handle = new StepHandle("step-1", "trace-1", "test-source", events::add);
    }

    @Test
    void getIds() {
        assertEquals("step-1", handle.getStepId());
        assertEquals("trace-1", handle.getTraceId());
    }

    @Test
    void finishSendsEvent() {
        handle.finish(Map.of("result", "done"));

        assertEquals(1, events.size());
        assertEquals(TraceEventType.STEP_FINISHED, events.get(0).eventType());
        assertEquals("step-1", events.get(0).stepId());
        assertEquals("trace-1", events.get(0).traceId());
    }

    @Test
    void finishNoArgs() {
        handle.finish();
        assertEquals(1, events.size());
        assertEquals(TraceEventType.STEP_FINISHED, events.get(0).eventType());
    }

    @Test
    void finishIgnoredWhenClosed() {
        handle.finish();
        handle.finish();
        assertEquals(1, events.size());
    }

    @Test
    void failWithString() {
        handle.fail("step error");

        assertEquals(1, events.size());
        assertEquals(TraceEventType.STEP_FAILED, events.get(0).eventType());
        assertEquals("step error", events.get(0).payload().get("error"));
    }

    @Test
    void failWithThrowable() {
        handle.fail(new RuntimeException("step exception"));

        assertEquals(1, events.size());
        assertEquals(TraceEventType.STEP_FAILED, events.get(0).eventType());
        assertEquals("step exception", events.get(0).payload().get("error"));
        assertNotNull(events.get(0).payload().get("stack"));
    }

    @Test
    void logSendsEvent() {
        handle.log("step log", LogLevel.ERROR, "step-event", Map.of("detail", 42));

        assertEquals(1, events.size());
        assertEquals(TraceEventType.LOG_EMITTED, events.get(0).eventType());
        assertEquals("step log", events.get(0).payload().get("message"));
        assertEquals("ERROR", events.get(0).payload().get("level"));
        assertEquals("step-1", events.get(0).stepId());
    }

    @Test
    void logDefaultLevel() {
        handle.log("default level");

        assertEquals("INFO", events.get(0).payload().get("level"));
    }

    @Test
    void isClosedReturnsFalseInitially() {
        assertFalse(handle.isClosed());
    }

    @Test
    void isClosedReturnsTrueAfterFinish() {
        handle.finish();
        assertTrue(handle.isClosed());
    }

    @Test
    void isClosedReturnsTrueAfterFail() {
        handle.fail("error");
        assertTrue(handle.isClosed());
    }

    @Test
    void closeAutoFailsIfNotClosed() {
        handle.close();

        assertTrue(handle.isClosed());
        assertEquals(1, events.size());
        assertEquals(TraceEventType.STEP_FAILED, events.get(0).eventType());
        assertEquals("Step not explicitly closed", events.get(0).payload().get("error"));
    }

    @Test
    void closeIsNoopIfAlreadyClosed() {
        handle.finish();
        handle.close(); // should not send another event

        assertEquals(1, events.size());
    }
}

File: php/laravel/tests/Unit/Handles/StepHandleTest.php
<?php

namespace Smartness\TraceFlow\Tests\Unit\Handles;

use PHPUnit\Framework\TestCase;
use Smartness\TraceFlow\DTO\TraceEvent;
use Smartness\TraceFlow\Enums\TraceEventType;
use Smartness\TraceFlow\Handles\StepHandle;

class StepHandleTest extends TestCase
{
    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private function makeStep(
        string $stepId = 'step-xyz',
        string $traceId = 'trace-abc',
        array &$events = [],
    ): StepHandle {
        $sendEvent = function (TraceEvent $event) use (&$events): void {
            $events[] = $event;
        };

        return new StepHandle(
            stepId: $stepId,
            traceId: $traceId,
            source: 'test',
            sendEvent: $sendEvent,
        );
    }

    // -------------------------------------------------------------------------
    // finish()
    // -------------------------------------------------------------------------

    public function test_finish_sends_step_finished_event(): void
    {
        $events = [];
        $step = $this->makeStep(events: $events);

        $step->finish('output-value', ['m' => 1]);

        $this->assertCount(1, $events);
        $this->assertSame(TraceEventType::STEP_FINISHED, $events[0]->eventType);
        $this->assertSame('output-value', $events[0]->payload['output']);
        $this->assertSame(['m' => 1], $events[0]->payload['metadata']);
        $this->assertSame('step-xyz', $events[0]->stepId);
    }

    public function test_finish_is_idempotent(): void
    {
        $events = [];
        $step = $this->makeStep(events: $events);

        $step->finish();
        $step->finish();

        $this->assertCount(1, $events);
    }

    // -------------------------------------------------------------------------
    // fail()
    // -------------------------------------------------------------------------

    public function test_fail_with_string(): void
    {
        $events = [];
        $step = $this->makeStep(events: $events);

        $step->fail('error msg');

        $this->assertCount(1, $events);
        $this->assertSame(TraceEventType::STEP_FAILED, $events[0]->eventType);
        $this->assertSame('error msg', $events[0]->payload['error']);
    }

    public function test_fail_with_throwable_captures_stack(): void
    {
        $events = [];
        $step = $this->makeStep(events: $events);

        $exception = new \RuntimeException('something went wrong');
        $step->fail($exception);

        $this->assertCount(1, $events);
        $this->assertSame(TraceEventType::STEP_FAILED, $events[0]->eventType);
        $this->assertSame('something went wrong', $events[0]->payload['error']);
        $this->assertNotEmpty($events[0]->payload['stack']);
    }

    // -------------------------------------------------------------------------
    // isClosed()
    // -------------------------------------------------------------------------

    public function test_is_closed_initially_false(): void
    {
        $events = [];
        $step = $this->makeStep(events: $events);

        $this->assertFalse($step->isClosed());
    }

    public function test_is_closed_true_after_finish(): void
    {
        $events = [];
        $step = $this->makeStep(events: $events);

        $step->finish();

        $this->assertTrue($step->isClosed());
    }

    public function test_is_closed_true_after_fail(): void
    {
        $events = [];
        $step = $this->makeStep(events: $events);

        $step->fail('x');

        $this->assertTrue($step->isClosed());
    }

    // -------------------------------------------------------------------------
    // log()
    // -------------------------------------------------------------------------

    public function test_log_sends_log_emitted_with_step_id(): void
    {
        $events = [];
        $step = $this->makeStep(stepId: 'step-xyz', events: $events);

        $step->log('msg', 'DEBUG', null, null);

        $this->assertCount(1, $events);
        $this->assertSame(TraceEventType::LOG_EMITTED, $events[0]->eventType);
        $this->assertSame('step-xyz', $events[0]->stepId);
    }

    // -------------------------------------------------------------------------
    // Destructor
    // -------------------------------------------------------------------------

    public function test_destructor_auto_closes_unclosed_step(): void
    {
        $events = [];
        $sendEvent = function (TraceEvent $event) use (&$events): void {
            $events[] = $event;
        };

        $step = new StepHandle(
            stepId: 'step-auto',
            traceId: 'trace-abc',
            source: 'test',
            sendEvent: $sendEvent,
        );

        // Destroy without calling finish/fail
        unset($step);

        $this->assertNotEmpty($events);
        $this->assertSame(TraceEventType::STEP_FAILED, $events[0]->eventType);
    }

    public function test_destructor_skips_already_closed_step(): void
    {
        $events = [];
        $sendEvent = function (TraceEvent $event) use (&$events): void {
            $events[] = $event;
        };

        $step = new StepHandle(
            stepId: 'step-closed',
            traceId: 'trace-abc',
            source: 'test',
            sendEvent: $sendEvent,
        );

        $step->finish('done');
        $countAfterFinish = count($events);

        // Destructor must not fire a second event
        unset($step);

        $this->assertCount($countAfterFinish, $events);
    }
}
